addons/website_sale_product_configurator/controllers/main.py

```diff
@@ -33,67 +34,54 @@ def _prepare_product_values(self, product, category, search, **kwargs):
+                'unique_id',
+                'product_custom_attribute_values',
+                'no_variant_attribute_values'
+            }
+        """
         if lang:
             request.website = request.website.with_context(lang=lang)
 
         order = request.website.sale_get_order(force_create=True)
         if order.state != 'draft':
             request.session['sale_order_id'] = None
             order = request.website.sale_get_order(force_create=True)
-        optional_product_ids = []
-        for k, v in kw.items():
-            if "optional-product-" in k and int(kw.get(k.replace("product", "add"))):
-                optional_product_ids.append(int(v))
 
-        custom_values = []
-        if kw.get('custom_values'):
-            custom_values = json.loads(kw.get('custom_values'))
+        custom_values = json.loads(custom_values)
+        if len(custom_values) != 0:
+            # the main product is the first, optional products are the rest
+            main_product = custom_values[0]
+            optional_products = custom_values[1:]
+            optional_product_ids = [p['product_id'] for p in optional_products]
 
-        value = {}
-        if add_qty or set_qty:
             value = order._cart_update(
-                product_id=int(product_id),
-                add_qty=add_qty,
-                set_qty=set_qty,
+                product_id=main_product['product_id'],
+                add_qty=main_product['quantity'],
                 optional_product_ids=optional_product_ids,
-                product_custom_attribute_values=self._get_product_custom_value(
-                    int(product_id),
-                    custom_values,
-                    'product_custom_attribute_values'
-                ),
-                no_variant_attribute_values=self._get_product_custom_value(
-                    int(product_id),
-                    custom_values,
-                    'no_variant_attribute_values'
-                )
+                product_custom_attribute_values=main_product['product_custom_attribute_values'],
+                no_variant_attribute_values=main_product['no_variant_attribute_values'],
             )
 
-        # options have all time the same quantity
-        for option_id in optional_product_ids:
-            order._cart_update(
-                product_id=option_id,
-                set_qty=value.get('quantity'),
-                linked_line_id=value.get('line_id'),
-                product_custom_attribute_values=self._get_product_custom_value(
-                    option_id,
-                    custom_values,
-                    'product_custom_attribute_values'
-                ),
-                no_variant_attribute_values=self._get_product_custom_value(
-                    option_id,
-                    custom_values,
-                    'no_variant_attribute_values'
+            # link option with its parent.
+            # Parents should always be listed before there options.
```

Comment:
I understand the concern about these constraint. My only goal here was to avoid browsing the array multiple times. I think it will not slow the process too much to reorder the array.